### PR TITLE
Change the installation steps, rework fame Dockerfile

### PR DESCRIPTION
**Pull request by Augustin-FL**
This pull requests does ~~3~~ 2 things : 

- Use environment variables for the setup steps for instead of `input()`. This way, the deployment can be easily automated
- ~~Unlock dependencies in `requirements.txt` : Flask is now in 2.0.x, some dependencies are now obsolete~~
- Rework the Dockerfile and improve the setup of a fame instance using docker.

### CODE REVIEW

**Review by certsocietegenerale — APPROVED** [review]

**Comment by certsocietegenerale** [comment]
@gaelmuller can we merge this? would you like to make a review?

File: utils/install.py [hunk 1]
```diff
@@ -4,13 +4,17 @@
...
 
 sys.path.append(os.path.normpath(os.path.join(os.path.dirname(os.path.abspath(__file__)), "..")))
 
-from utils import user_input, error
+from utils import error
 from fame.common.constants import FAME_ROOT
 from fame.common.pip import pip_install
 
+env = Env()
+env.read_env(env.str('ENV_PATH', '.env'), recurse=False)
```
Comment (gaelmuller) [inline utils/install.py:16]: Why use a third party library to read environment variables ?

File: utils/run.sh [hunk 2]
```diff
@@ -6,12 +6,19 @@ VIRTUALENV="$FAME_ROOT/env"
 
 if [ -d "$VIRTUALENV" ]; then
     echo "[+] Using existing virtualenv."
+    . "$VIRTUALENV/bin/activate"
 else
     echo "[+] Creating virtualenv..."
     python3 -m virtualenv -p python3 "$VIRTUALENV" > /dev/null
+    . "$VIRTUALENV/bin/activate"
+    pip3 install -r "$FAME_ROOT"/requirements.txt
```
Comment (gaelmuller) [inline utils/run.sh:14]: Why this change ? The installation script is already installing the requirements.

File: web/views/analyses.py [hunk 3]
```diff
@@ -72,7 +72,7 @@ def index(self):
         page = int(request.args.get('page', 1))
 
         analyses = current_user.analyses.find().sort('_id', DESCENDING).limit(PER_PAGE).skip((page - 1) * PER_PAGE)
-        pagination = Pagination(page=page, per_page=PER_PAGE, total=analyses.count(), css_framework='bootstrap3')
+        pagination = Pagination(page=page, per_page=PER_PAGE, total=len(list(analyses.clone())), css_framework='bootstrap3')
```
Comment (gaelmuller) [inline web/views/analyses.py:75]: In my understanding this is not the same thing.

- `.count()` will ignore the limit and return the count of all results
- Your proposal will not ignore the limit and will perform the operation twice (once for the count and once to actually get the analyses)

The same goes for all `count` replacements.

`count_documents` appears to be the correct replacement, but not as easy to use. We could also chose to change how pagination is handled because having a count is not that useful anyway.

**Review by gaelmuller — CHANGES_REQUESTED** [review]
Thank you for your contribution !

My main concern is with the dependency updates and associated changes. IMHO we should always specify the versions to use to avoid incompatibility issues.

File: utils/run.sh (hunk 2 above)
Comment (Augustin-FL) [reply utils/run.sh:14]: The idea here is to install python requirements when performing `docker build .`, instead of when launching the container.

In order to do this, I had to 
- update utils/run.sh to install requirements when creating the environment
- and run utils/run.sh in the Dockerfile itself

I still kept the requirements.txt installation in the .py, in order to not disturb non-docker installations

File: utils/run.sh [hunk 4]
```diff
@@ -6,12 +6,19 @@ VIRTUALENV="$FAME_ROOT/env"
...
-. "$VIRTUALENV/bin/activate"
 
 echo ""
-python "$@"
+if [ -n "$*" ]; then
+    progname=$1
+    shift
+    args="$*"
+    python "$progname" "$args"
+fi
```
[the remark this answers is not in the document]
Comment (Augustin-FL) [reply utils/run.sh:24]: the idea here is to launch python with arguments in order to automate the setup

`python utils/install.py worker` will install a remote worker
`python utils/install.py web` will install a fame web server

I thought it was clearer this way.  I also took care of if utils/run.sh was run without arguments (in this case, no python launch is needed? If you want to launch a python interpreter there is already shell.py for this purpose, if I' mot mistaken)

File: web/views/analyses.py (hunk 3 above)
Comment (Augustin-FL) [reply web/views/analyses.py:75]: Hum...
Thanks for bringing this to me, I missed the limit part.

I agree this code is a bit dirty: I was trying to avoid changing how the pagination is done entirely, and as you said `count_documents()` is not as easy to use. I will rework this and come back to you with a new proposal

File: utils/install.py (hunk 1 above)
Comment (Augustin-FL) [reply utils/install.py:16]: Because env library easily handle default values. 

`context['fame_url'] = env("FAME_URL", "http://localhost")`

If you want, I can also remove this lib and use a lot of "if" instead
```
context['fame_url'] =  'http://localhost'
if 'FAME_URL' in os.environ and os.environ['FAME_URL']:
    context['fame_url'] =  os.environ['FAME_URL']
```

would you like me to do that?

File: docker/Dockerfile [hunk 5]
```diff
@@ -1,45 +1,45 @@
...
-  python-pip \
+  python3-dev \
+  python3-pip \
   screen \
   p7zip-full \
   libjpeg-dev \
   zlib1g-dev \
+  libmagic-dev \
   apt-transport-https \
+  netcat \
```
[the remark this answers is not in the document]
Comment (Augustin-FL) [reply docker/Dockerfile:19]: yes, for `wait-for`

File: docker/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    env_file:
+      - fame.env
+    networks:
+      - backend.fame
+    environment:
+      MONGO_INITDB_ROOT_USERNAME: admin
+      MONGO_INITDB_ROOT_PASSWORD: NeedToBeChanged
+    volumes:
+      - mongodb:/data/db
+      - ./mongo-init-root-user.js:/docker-entrypoint-initdb.d/mongo-init-root-user.js:ro
```
[the remark this answers is not in the document]
Comment (Augustin-FL) [reply docker/docker-compose.yml:62]: This is needed to give the fame user limited privileges on mongodb. 
There is no easy way to setup a root mongodb account+a secondary account with limited privileges for the app. The docker environment variables (`MONGO_INITDB_ROOT_*`) can only setup a root user, as the name suggests.

The solution I came up with comes mostly from https://stackoverflow.com/questions/42912755/how-to-create-a-db-for-mongodb-container-on-start-up . I think it's an acceptable one ?

if you want, the mongodb setup can be moved elsewhere (`install.py`?) or removed entirely. Could you confirm if you wish me to do that?

File: docker/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    networks:
+      - backend.fame
+    ports:
+      - "4200:4200"
+    env_file:
+      - fame.env
+    depends_on:
+      - fame_db
+    volumes:
+      - modules:/fame/fame/modules/
```
[the remark this answers is not in the document]
Comment (Augustin-FL) [reply docker/docker-compose.yml]: You are entirely right. I realize I seriously messed up here. Thanks for casting some light on it

File: requirements.txt [hunk 8]
```diff
@@ -1,14 +1,18 @@
...
-sphinx_rtd_theme~=0.5.0
-sphinxcontrib-httpdomain~=1.7.0
-celery~=4.4.7
-pymongo~=3.11.0
-python-magic~=0.4.18
-requests~=2.24.0
-Flask-Login==0.5.0
-zxcvbn==4.4.28
-GitPython~=3.1.7
-markdown2~=2.3.9
```
[the remark this answers is not in the document]
Comment (Augustin-FL) [reply requirements.txt:14]: I changed this. version are now locked again.

I wanted to unlock the dependencies because I thought having "up to date" packages was more important than preventing stability issues. From your comment, I understand that you disagree 😅

**Comment by Augustin-FL** [comment]
Hi @gaelmuller,

Sorry for the delay. would it be possible to request a second review?

File: utils/install.py (hunk 1 above)
Comment (gaelmuller) [reply utils/install.py]: You can actually do the same without a library:

```python
os.getenv("FAME_URL", "http://localhost")
```

or

```python
os.environ.get("FAME_URL", "http://localhost")
```

File: utils/run.sh (hunk 2 above)
Comment (gaelmuller) [reply utils/run.sh:14]: non-docker installations are also using `utils/run.sh` so it would probably make more sense to move requirements installation in `run.sh` in all cases.

File: utils/run.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (gaelmuller) [reply utils/run.sh]: I may be missing something but in my understanding this is already what `$@` was doing. Maybe we need to remove the quotes ?

Anyway I'm fine with both versions

File: docker/docker-compose.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaelmuller) [reply docker/docker-compose.yml:66]: That's fine :+1:

File: docs/installation.rst [hunk 9]
```diff
@@ -5,6 +5,26 @@ Installation
...
 
+The easy way
+============
+
+Probably the most easy way to run fame is::
+
+    $ git clone https://github.com/certsocietegenerale/fame.git
+    $ cd fame/docker
+    $ cp fame.env.template fame.env
+    $ sed -i "s/#privileged: true/privileged: true/g" docker-compose.yml
```
Comment (gaelmuller) [inline docs/installation.rst:16]: Could you explain why you the change from Docker out of Docker to Docker in Docker ? I do not have experience with any of them but the first one seemed good to me.

If we do chose to keep this Docker in Docker approach, we should explain that this step is optional and when it is necessary. Otherwise, we should directly uncomment it in the docker-compose file.

File: docs/installation.rst [hunk 10]
```diff
@@ -5,6 +5,26 @@ Installation
...
+    $ docker-compose up --build
+
+
+Then browse http://localhost:4200/ using your web browser, and login using the `docker first run default credentials <https://github.com/certsocietegenerale/fame/blob/master/docker/fame.env.template>`_  (admin@changeme.fame / changeme)
+
+These commands are very nice for having a first look to what fame is, but you may also want to install fame without docker.
+
+This is especially true given that FAME use docker internally for its activities, which leads to Docker-in-Docker situation. DinD is dangerous, we do not recommend running FAME in docker in production.
```
Comment (gaelmuller) [inline docs/installation.rst:24]: I'm not sure this is true. Is it really more dangerous than a typical installation where the fame users is in the docker users ?

File: web/views/analyses.py [hunk 11]
```diff
@@ -72,7 +72,7 @@ def index(self):
         page = int(request.args.get('page', 1))
 
         analyses = current_user.analyses.find().sort('_id', DESCENDING).limit(PER_PAGE).skip((page - 1) * PER_PAGE)
-        pagination = Pagination(page=page, per_page=PER_PAGE, total=analyses.count(), css_framework='bootstrap3')
+        pagination = Pagination(page=page, per_page=PER_PAGE, total=current_user.analyses.estimated_document_count(), css_framework='bootstrap3')
```
Comment (gaelmuller) [inline web/views/analyses.py:75]: Unfortunately this is still not the same. This would return the count for the whole collection instead of with the applied filters.

File: docker/docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    env_file:
+      - fame.env
+    depends_on:
+      - fame_db
+    volumes:
+      - modules:/opt/fame/fame/modules/
+      - storage:/opt/fame/fame/storage/
+      - avatars:/opt/fame/web/static/img/avatars/
+
+  fame_worker:
```
Comment (gaelmuller) [inline docker/docker-compose.yml:33]: I would expect the worker to have the modules and storage volumes since it is not a "remote" worker.

**Review by gaelmuller — CHANGES_REQUESTED** [review]

File: docs/installation.rst (hunk 10 above)
Comment (Augustin-FL) [reply docs/installation.rst:24]: You are absolutely right.
I wrote this as an abundance of caution : many people have a false sense that "docker containers are like virtual machines : they are isolated sandboxes". This is true to some extend but...container evasion is a lot more easier than VM evasion. Especially when the container is privileged.

I wrote this disclaimer because I wanted to clarify that fame worker really need to be on its dedicated server. I will reword this to avoid spreading FUD

File: docs/installation.rst (hunk 9 above)
Comment (Augustin-FL) [reply docs/installation.rst:16]: In general, DooD is recommended in most cases : Unlike DinD, DooD does not require privileged containers, and this limit (a little bit) the possibilities for container evasion.
Note that container evasion is still possible with DooD (since we have access to the docker socket, we can spawn a privileged container on the host...) but It is considered cleaner to perform DooD anyway

Why did I switch to DinD, then?
Because of one of the limitations with DooD : [bind volumes between parent docker and child are very complex](http://ralphmcneal.com/dood-pitfalls-volume-mounts/), and dynamic volume creation is not supported (all volumes must be created before starting the parent docker)

Bind volumes are being used to transfer data between FAME modules and a fame worker instance. I could have reworked how fame modules transfers are being done (in fact, that's what https://github.com/certsocietegenerale/fame_modules/pull/27 tried to do, if I'm not mistaken) but I instead opted out to switch to DinD

About `sed` specifically:
If you don't execute the sed command, then modules using docker (url_preview, etc) won't work. Modules which are not using docker will work just fine
I made it this way to "force" users to understand the container is privileged. 

Do you confirm that I should uncomment it directly in the docker-compose?

File: docker/docker-compose.yml (hunk 12 above)
Comment (Augustin-FL) [reply docker/docker-compose.yml:33]: Actually, `fame_worker` is a remote worker, not a local one. It's a separate docker container, with a dedicated IP. This container can only communicate with `fame_web` and with the db

File: docs/installation.rst (hunk 9 above)
Comment (Augustin-FL) [reply docs/installation.rst]: I removed the `sed` part in the doc. I'm assuming this can be marked as resolved

File: utils/run.sh (hunk 2 above)
Comment (Augustin-FL) [reply utils/run.sh:14]: Done

**Comment by Augustin-FL** [comment]
Thanks for your second review (and sorry for the stupid mistake I made... `estimated_document_count`, etc)

I'm assuming this time it should be good for merge?

File: docker/docker-compose.yml (hunk 12 above)
Comment (gaelmuller) [reply docker/docker-compose.yml:33]: Alright, I missed that but it will not work.

The `updates` queue that is supported by this worker is responsible for everything related to module repository management. With current code, a "local" worker is needed to support the `updates` queue

Comment (Augustin-FL) [reply docker/docker-compose.yml:33]: I fixed the problem : there is now an additional container `fame_updater` which has modules and storage volumes mounted, and behave like a local worker

File: docker/Dockerfile [hunk 13]
```diff
@@ -1,45 +1,57 @@
...
-    mongodb-org
+RUN groupadd -r fame && \
+    useradd -M -r -d /opt/fame -g fame -G docker fame && \
+    install -d -m 0755 -o fame -g fame /opt/fame
+
+USER fame
+
+RUN git clone https://github.com/certsocietegenerale/fame && \
+    pip3 install virtualenv && \
+    /opt/fame/utils/run.sh
```
Comment (gaelmuller) [inline docker/Dockerfile:50]: I would use `COPY . /opt/fame` instead of cloning the repository. The docker deployment is most likely to be used as a development version.

File: docker/Dockerfile [hunk 14]
```diff
@@ -1,45 +1,57 @@
...
 
+USER root
```
Comment (gaelmuller) [inline docker/Dockerfile:52]: Why run things as `root` and then `sudo` to `fame` ?

File: docker/Dockerfile [hunk 15]
```diff
@@ -1,45 +1,57 @@
...
 
-RUN pip install virtualenv && \
-    git clone https://github.com/certsocietegenerale/fame
+RUN chown -R fame:fame /opt/fame
```
Comment (gaelmuller) [inline docker/Dockerfile:54]: I don't think this is needed

File: docker/docker-compose.yml [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+    networks:
+      - backend.fame
+    env_file:
+      - fame.env
+    privileged: true # this is required to use docker-in-docker functionnalities. Without privilegied mode, FAME modules which are using docker will not work
+    depends_on:
+      - fame_db
+      - fame_web
+
+  fame_updater:
```
Comment (gaelmuller) [inline docker/docker-compose.yml:51]: I don't really get the point of having two workers in this deployment. I would expect only one local worker, but two is fine.

File: docker/docker-compose.yml [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    env_file:
+      - fame.env
+    networks:
+      - backend.fame
+    environment:
+      MONGO_INITDB_ROOT_USERNAME: admin
+      MONGO_INITDB_ROOT_PASSWORD: NeedToBeChanged
+    volumes:
+      - mongodb:/data/db
+      - ./mongo-init-root-user.js:/docker-entrypoint-initdb.d/mongo-init-root-user.js:ro
```
Comment (gaelmuller) [inline docker/docker-compose.yml:85]: This is not working for me because `_getEnv` does not exist on latest mongodb versions

**Review by gaelmuller — CHANGES_REQUESTED** [review]
How did you test this ? :thinking: 

This is not working on my end (because FAME cannot connect to the database)

**Comment by Augustin-FL** [comment]
> How did you test this ? 🤔
> 
> This is not working on my end (because FAME cannot connect to the database)

Like this : 

```
git clone https://github.com/Augustin-FL/fame
cd fame/docker
git checkout use_env

sed -i 's+git clone https://github.com/certsocietegenerale/fame+git clone https://github.com/Augustin-FL/fame \&\& cd fame \&\& git checkout use_env \&\& cd ..+g' Dockerfile 

cp fame.env.template fame.env
docker-compose build
docker-compose up
```

Alternatively, you could indeed use `COPY` instead of `RUN git clone`, like you did.


What do you mean "FAME cannot connect to the database" ? It does work well on my side ?

docker-compose

File: docker/docker-compose.yml (hunk 16 above)
Comment (Augustin-FL) [reply docker/docker-compose.yml:51]: Well, I just tried to respect the docker philosophy here 😄 

Ideally, each docker should be used for one task only. This push you to "break down" a monolithic application into multiple "microservices". 

That being said : if you want I can indeed move back the update function to the web server. Could you confirm you would like me to do that?

File: docker/Dockerfile (hunk 14 above)
Comment (Augustin-FL) [reply docker/Dockerfile:52]: Because : 
- In general, everything should run as `fame` user
- however, there is one command I need to run as root in the entry point : `dockerd &`

https://github.com/certsocietegenerale/fame/blob/4cdfd4d034709d3a4902adc0b186a2e0d8afaf7c/docker/docker-entrypoint.sh#L7

I could have run everything as fame and add a sudoers entry, but I was lazy and I instead decided to run as root and demote to fame. Not a big deal ?

File: docker/Dockerfile (hunk 15 above)
Comment (Augustin-FL) [reply docker/Dockerfile:54]: It is required because the web server & worker are being run as `fame` user
The web server/workers need to create files/folders (temp, storage, users avatars, etc), so a chown is required

of course, this is not required anymore if we decide to run as root

File: docker/Dockerfile (hunk 13 above)
Comment (Augustin-FL) [reply docker/Dockerfile:50]: Hum

Okay, make sense

**Comment by Augustin-FL** [comment]
> What do you mean "FAME cannot connect to the database" ? It does work well on my side ?

Hi again,

I understand what do you mean now. the issue is the following : mongodb unexpectedly quit after performing init process.
I didn't experience the issue because I often `CTRL+C` when doing my builds, and I don't always reset my volumes.

Let me fix this

**Comment by Augustin-FL** [comment]
I think you can make a (hopefully final?) review now
